### PR TITLE
[BREAKING CHANGE] Major refactor for v4.0

### PR DESCRIPTION
**Pull request by TangoMan75**
Hi Arthur,

Sorry for fat pull request but since I pushed backward compatibility breaking changes anyway, I squashed all changes in one commit.

Here's the list of changes I made:

New tools
---------

- `NetscapeBookmarkDecoder` and `NetscapeBookmarkEncoder`.
- `BookmarksDataFormatter` class to use before export to have bookmarks ordered by tag.
- `BookmarksDataConverter` and `LegacyBookmarksDataConverter` class to convert old bookmarks to new data format.

Refactor `parseBoolean`
-----------------------

- `parseBoolean` should only accept strings as entry parameter.
- `parseBoolean` should always return boolean.
- `$item['public']` should either contain `true`, `false` or `null` (when undefined);
- Removed integers, data comes from a text file ; It will simply just never contain integer value.

Deprecations
------------

Tagged most methods from `NetscapeBookmarkParser` for deprecation.

Should be remove in the next major release.

Refactor constructor
--------------------

> Replace parameters from constructor with "$context" array and class constants.
> This will allow for Open/Close principle enforcement

Undefined values should remain `null`
-------------------------------------

This is not the decoder job to set default values.

### Undefined Title should return `null`

> decoder should not set default `$item['name'] = 'untitled'`.

### Undefined Time should return `null`

> decoder should not set `$item['dateCreated'] = time()` as default.

### Undefined Note should return `null`

> A note equals to an empty string should return `null`.

Remove default Tags
-------------------

> Default Tags should be removed ; This is not the decoder job to set default values.

Remove default public/private status
------------------------------------

> Default public/private status should be removed ; This is not the decoder job to set default values.
> Undefined `$item['public']` should return `null`

Replaced Travis CI with GitHub actions
--------------------------------------

> This allow any contibutor to have the ci pipeline available within their own fork.

Refactor Tests
--------------

Full factorisation of Unit tests.

Updated minimum required version
--------------------------------

https://www.php.net/supported-versions.php

php 7.1 and 7.2 have reached end of life.

```json
// composer.json
  "require": {
    "php": ">=7.3"
  },
```
Note that, php 7.3 will be supported until december 2021.

Updated PHPUnit version
-----------------------

Updated PHPUnit to the latest version.

```json
  "require-dev": {
    // ...
    "phpunit/phpunit": "^9.5"
  }
```

### CODE REVIEW

**Comment by TangoMan75** [comment]
Hi Arthur,

I understand this is a lot to process all at once, I hope you find the time to merge this PR.

This was actually quite a lot of work, I would be really happy to get your feedback on this.

Regards

File: .editorconfig [hunk 1]
```diff
@@ -1,19 +0,0 @@
-# EditorConfig: http://EditorConfig.org
```
Comment (ArthurHoaro) [inline .editorconfig:1]: Why did you remove this file?

File: .gitattributes [hunk 2]
```diff
@@ -1,14 +0,0 @@
-# Set default behavior
```
Comment (ArthurHoaro) [inline .gitattributes:1]: Why did you remove this file?

File: composer.json [hunk 3]
```diff
@@ -15,21 +15,31 @@
...
+    },
     "autoload-dev": {
         "psr-4": {
-            "Shaarli\\NetscapeBookmarkParser\\": "tests/"
+            "Shaarli\\NetscapeBookmarkParser\\Tests\\": "tests/"
         }
     },
     "require": {
-        "php": ">=7.1",
+        "php": ">=7.3",
```
Comment (ArthurHoaro) [inline composer.json]: While I agree that PHP < 7.3 has reached EoL, this project is still tied to Shaarli which support PHP 7.1+. Shaarli's PHP version requirement will be bumped eventually, but for now I'd rather not make this new version incompatible.

Also, there is not a lot of syntax changes between 7.1 and 7.3.

File: src/DataConverter/DataConverterInterface.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    /**
+     * Convert source array to destination format.
+     */
+    public function convert(array $source): array;
+
+    /**
+     * Reverse convert back to original array.
+     */
+    public function reverseConvert(array $source): array;
```
Comment (ArthurHoaro) [inline src/DataConverter/DataConverterInterface.php]: Since you are declaring an interface widely used in the project here, having a more explicit PHPDoc about expected parameters would be helpful, especially if you can provide an example.

File: src/DataConverter/AbstractConverter.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+namespace Shaarli\NetscapeBookmarkParser\DataConverter;
+
+/**
+ * @author "Matthias Morin" <mat@tangoman.io>
+ */
+abstract class AbstractConverter implements DataConverterInterface
+{
+    public const FORMAT = '';
```
Comment (ArthurHoaro) [inline src/DataConverter/AbstractConverter.php]: Just to avoid having a valid empty `FORMAT` in sub classes:

```suggestion
    public const FORMAT = null;
```

File: src/DataConverter/BookmarkConverter.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Shaarli\NetscapeBookmarkParser\DataConverter;
+
+/**
+ * @author "Matthias Morin" <mat@tangoman.io>
+ */
+class BookmarkConverter extends AbstractConverter
```
Comment (ArthurHoaro) [inline src/DataConverter/BookmarkConverter.php:21]: A comment giving the context of this type would help. I don't really know what I'm looking at.

File: src/DataConverter/LegacyBookmarkConverter.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Shaarli\NetscapeBookmarkParser\DataConverter;
+
+/**
+ * @author "Matthias Morin" <mat@tangoman.io>
+ */
+class LegacyBookmarkConverter extends AbstractConverter
```
Comment (ArthurHoaro) [inline src/DataConverter/LegacyBookmarkConverter.php:21]: A comment giving the context of this type would help. I don't really know what I'm looking at.

File: src/DataFormatter/BookmarksDataFormatter.php [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+        $tags = [];
+        foreach ($data as $bookmark) {
+            foreach ($bookmark['tags'] ?? [] as $tagName) {
+                $tags[$tagName][] = $bookmark;
+            }
+        }
+
+        $data = [];
+        foreach ($tags as $tagName => $bookmarks) {
```
Comment (ArthurHoaro) [inline src/DataFormatter/BookmarksDataFormatter.php:59]: If a bookmark doesn't have tags, it is lost here. Is it an expected behavior?

**Comment by ArthurHoaro** [comment]
Hi Mattias, I understand it's frustrating to put a lot of effort into some work and not get any feedback. I'm sorry about that, I just don't really have time for Shaarli these days.

I'm just letting you know that I have started to review this PR, and it looks great! Awesome work!
I'll continue this week, because it's pretty massive, and will have a few comments but overall there shouldn't be too much changes. Especially with our test dataset: if previous UT are still passing we can be confident about these changes.

However you did quite some changes about this repo metadata that weren't discussed beforehand, and it would be nice if you could give more context about all of them (code of conduct, etc.).

**Comment by TangoMan75** [comment]
Hi Arthur, I understand you've been busy, no problem.

> However you did quite some changes about this repo metadata that weren't discussed beforehand, and it would be nice if you could give more context about all of them (code of conduct, etc.).

Yeah, I know... Sorry about that... I got carried away a bit XD

```
CODE_OF_CONDUCT.md
ISSUE_TEMPLATE/bug_report.md
ISSUE_TEMPLATE/feature_request.md
```

These are just template text files, we can change them any fashion you want.

I really love new GitHub Action feature, this allow contributors to have the ci pipeline available within their own fork with is a huge benefit for the project... So I removed Travis.

Also I bumped php and phpunit versions...

We can discuss any changes, take your time.

File: README.md [hunk 9]
```diff
@@ -1,108 +1,242 @@
-# netscape-bookmark-parser
-[![](https://img.shields.io/travis/shaarli/netscape-bookmark-parser/master.svg?style=flat-square&label=master)](https://travis-ci.org/shaarli/netscape-bookmark-parser)
-[![](https://img.shields.io/github/release/shaarli/netscape-bookmark-parser.svg?style=flat-square)](https://github.com/shaarli/netscape-bookmark-parser/releases/latest/)
-[![license](https://img.shields.io/github/license/shaarli/netscape-bookmark-parser.svg?style=flat-square)](https://opensource.org/licenses/MIT)
+![GH language](https://img.shields.io/github/languages/top/shaarli/netscape-bookmark-parser)
```
Comment (ArthurHoaro) [inline README.md]: This tag is pretty useless.

File: README.md [hunk 10]
```diff
@@ -1,108 +1,242 @@
...
+[![GH tag](https://img.shields.io/github/v/tag/shaarli/netscape-bookmark-parser)](https://github.com/shaarli/netscape-bookmark-parser/tags)
+[![GH release](https://img.shields.io/github/v/release/shaarli/netscape-bookmark-parser)](https://github.com/shaarli/netscape-bookmark-parser/releases)
```
Comment (ArthurHoaro) [inline README.md]: Those are usually the same.

File: README.md [hunk 11]
```diff
@@ -1,108 +1,242 @@
...
+$ php -r "copy('https://getcomposer.org/composer-stable.phar', 'composer.phar');"
+# install composer globally
+$ sudo mv composer.phar /usr/local/bin/composer
+# fix permissions
+$ sudo chmod uga+x /usr/local/bin/composer
+$ sync
+# install symfony flex globally to speed up download of composer packages (parallelized prefetching)
+$ composer global require 'symfony/flex' --prefer-dist --no-progress --no-suggest --classmap-authoritative
+$ composer clear-cache
+```
```
Comment (ArthurHoaro) [inline README.md]: I don't think it really makes sense to add how to install PHP and Composer here. It doesn't cover all cases, and may quickly become outdated. We should link to proper documentation instead.

File: README.md [hunk 12]
```diff
@@ -1,108 +1,242 @@
...
+
+Copyrights (c) 2021 &quot;Shaarli community&quot;
+
+[![License](https://img.shields.io/badge/Licence-MIT-green.svg)](LICENSE)
+Distributed under the MIT license.
+
+If you like **Shaarli Netscape Bookmark Parser** please star, follow or tweet about it:
+
+[![GitHub stars](https://img.shields.io/github/stars/shaarli/netscape-bookmark-parser?style=social)](https://github.com/shaarli/netscape-bookmark-parser/stargazers)
+[![GitHub followers](https://img.shields.io/github/followers/shaarli?style=social)](https://github.com/shaarli)
```
Comment (ArthurHoaro) [inline README.md]: I don't think it's possible to follow an organization on Github :D

File: src/DataFormatter/BookmarksDataFormatter.php [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+     *
+     * @return bool
+     */
+    public function supportsFormatting(string $format): bool
+    {
+        return self::FORMAT === $format;
+    }
+
+    /**
+     * Rearrange bookmarks array by tag name.
```
Comment (ArthurHoaro) [inline src/DataFormatter/BookmarksDataFormatter.php]: ~~I don't really understand in which context this is needed. It is used for export, but why would we want to group bookmarks by tag when exporting bookmarks into a netscape file?~~

OK I get it, that's how the Netscape file can be structured, grouped by tags. Maybe we could add a little comment here or in the encoder to explain that.

File: src/Encoder/NetscapeBookmarkEncoder.php [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        if (!\is_array($data)) {
+            throw new \InvalidArgumentException(
+                sprintf(
+                    'Argument 1 passed to %s::%s must be of type array, %s given',
+                    self::class,
+                    __FUNCTION__,
+                    \gettype($data)
+                )
+            );
+        }
```
Comment (ArthurHoaro) [inline src/Encoder/NetscapeBookmarkEncoder.php:60]: Why don't you type `$data` as an array instead?

File: src/NetscapeBookmarkParser.php [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+     * @return array An associative array containing parsed links
+     */
+    public function parseFile(string $filePath): array
+    {
+        $this->logger->info('Starting to parse ' . $filePath);
+
+        try {
+            return $this->decoder->decode(file_get_contents($filePath));
+        } catch (\Exception $exception) {
+            $this->logger->error($exception);
```
Comment (ArthurHoaro) [inline src/NetscapeBookmarkParser.php]: It should throw an exception, eventually a custom one. Let's stop silencing errors :D

File: src/NetscapeBookmarkParser.php [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+    {
+        $this->logger->info('Export to ' . $filePath);
+
+        $data = $this->formatter->format($data);
+
+
+        try {
+            $data = $this->encoder->encode($data);
+            var_dump($data);
+            die();
```
Comment (ArthurHoaro) [inline src/NetscapeBookmarkParser.php]: :grimacing: 

```suggestion
```

File: src/NetscapeBookmarkParser.php [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+
+    /**
+     * Parses a string containing Netscape-formatted bookmarks
+     *
+     * @param string $data
+     *
+     * @return array
+     * @deprecated
+     */
+    public function parseString(string $data): array
```
Comment (ArthurHoaro) [inline src/NetscapeBookmarkParser.php:111]: We should keep this one. Sometimes we don't want to deal with files and just manipulate strings.

File: src/NetscapeBookmarkParser.php [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+
+    /**
+     * Exports data to Netscape bookmark format file
+     *
+     * @param array  $data
+     * @param string $filePath
+     *
+     * @return string A string containing serialized data
+     */
+    public function export(array $data, string $filePath): string
```
Comment (ArthurHoaro) [inline src/NetscapeBookmarkParser.php:90]: We could also add a public method to export it as string without file writing operations.

File: src/NetscapeBookmarkParser.php [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+            return $this->decoder->decode($data);
+        } catch (\Exception $exception) {
+            $this->logger->error($exception);
+        }
+
+        return [];
+    }
+
+    /**
+     * @deprecated
```
Comment (ArthurHoaro) [inline src/NetscapeBookmarkParser.php:117]: I'm not sure it's necessary to keep those, as the next released version will be 4.0.

File: src/NetscapeBookmarkParser.php [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+     * @var BookmarksDataFormatter
+     */
+    private $formatter;
+
+    /**
+     * @var NetscapeBookmarkEncoder
+     */
+    private $encoder;
+
+    public function __construct(array $defaultContext = [], LoggerInterface $logger = null)
```
Comment (ArthurHoaro) [inline src/NetscapeBookmarkParser.php:57]: This is the entry point of the library. Context should be documented here.

File: src/NetscapeBookmarkParser.php (hunk 18 above)
Comment (ArthurHoaro) [reply src/NetscapeBookmarkParser.php:90]: Also this method is not covered by unit tests.

File: phpunit.xml.dist [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+<!-- https://phpunit.readthedocs.io/en/latest/configuration.html -->
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:noNamespaceSchemaLocation="vendor/phpunit/phpunit/phpunit.xsd"
+         convertDeprecationsToExceptions="true"
+         convertErrorsToExceptions="true"
+         convertNoticesToExceptions="true"
+         convertWarningsToExceptions="true"
+         colors="true"
+>
+    <testsuites>
```
Comment (ArthurHoaro) [inline phpunit.xml.dist:11]: Suggestion, include this for code coverage:

```suggestion
    <coverage>
        <include>
            <directory suffix=".php">./src</directory>
        </include>
    </coverage>
    <testsuites>
```

File: tests/Unit/Encoder/ParseChromiumBookmarksTest.php [hunk 22]
```diff
@@ -0,0 +1,349 @@
...
+            'url'         => 'https://owncloud.org/',
+            'tags'        => [
+                'self-hosting',
+            ],
+            'description' => null,
+            'dateCreated' => 1466013448,
+            'public'      => null,
+        ],
+    ];
```
Comment (ArthurHoaro) [inline tests/Unit/Encoder/ParseChromiumBookmarksTest.php:306]: This part must have been real fun :scream_cat:

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+A clear and concise description of what you expected to happen.
+
+**Screenshots**
+If applicable, add screenshots to help explain your problem.
+
+**Desktop (please complete the following information):**
+ - OS: [e.g. MacOS]
+ - make version (make --version)
+ - Script version
+ - copy/paste output
```
Comment (ArthurHoaro) [inline .github/ISSUE_TEMPLATE/bug_report.md]: This part is not really applicable for our library.

File: .github/workflows/php.yaml [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  php:
+    name: PHP project (PHP ${{ matrix.php-versions }})
+
+    runs-on: ubuntu-latest
+    if: "! contains(toJSON(github.event.commits.*.message), '[skip ci]')"
```
Comment (ArthurHoaro) [inline .github/workflows/php.yaml]: :eyes:

**Review by ArthurHoaro — CHANGES_REQUESTED** [review]
Impressive work @Tangoman75! :clap: :clap: 
I have added a few comments here and there, but it's mainly just great. Congrats.

There is only the PHP version bump that might be tricky to change.

**Comment by TangoMan75** [comment]
Thanks for your feedback @ArthurHoaro !

I'll get back to you soon.

File: .github/workflows/php.yaml (hunk 24 above)
Comment (TangoMan75) [reply .github/workflows/php.yaml]: I agree... This is suspicious... I'll remove it

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 23 above)
Comment (TangoMan75) [reply .github/ISSUE_TEMPLATE/bug_report.md]: I'll remove it...

File: README.md (hunk 9 above)
Comment (TangoMan75) [reply README.md]: ok, I'll remove it

File: README.md (hunk 10 above)
Comment (TangoMan75) [reply README.md]: Hmm... Maybe...

File: README.md (hunk 11 above)
Comment (TangoMan75) [reply README.md]: You're probably right

File: README.md (hunk 12 above)
Comment (TangoMan75) [reply README.md]: Haha, true... I'll remove it

File: composer.json (hunk 3 above)
Comment (TangoMan75) [reply composer.json]: Oh ! Alright... I'll change it back...

File: phpunit.xml.dist (hunk 21 above)
Comment (TangoMan75) [reply phpunit.xml.dist:11]: Good idea... Thanks

File: src/DataConverter/BookmarkConverter.php (hunk 6 above)
Comment (TangoMan75) [reply src/DataConverter/BookmarkConverter.php:21]: Sure

File: src/DataConverter/DataConverterInterface.php (hunk 4 above)
Comment (TangoMan75) [reply src/DataConverter/DataConverterInterface.php]: True... I believe linter removed it by accident...

File: src/DataConverter/LegacyBookmarkConverter.php (hunk 7 above)
Comment (TangoMan75) [reply src/DataConverter/LegacyBookmarkConverter.php:21]: ok

File: src/DataFormatter/BookmarksDataFormatter.php (hunk 13 above)
Comment (TangoMan75) [reply src/DataFormatter/BookmarksDataFormatter.php]: You're right... I'll do it

File: src/DataFormatter/BookmarksDataFormatter.php (hunk 8 above)
Comment (TangoMan75) [reply src/DataFormatter/BookmarksDataFormatter.php:59]: Wow ! I totally missed that ! This is NOT expected behavior ! I will need to fix this.

File: src/Encoder/NetscapeBookmarkEncoder.php (hunk 14 above)
Comment (TangoMan75) [reply src/Encoder/NetscapeBookmarkEncoder.php:60]: Well, I could but I want to have this class compatible with Symfony encoder interface in the future... Hmmm... Should I forget about it ?

File: src/NetscapeBookmarkParser.php (hunk 16 above)
Comment (TangoMan75) [reply src/NetscapeBookmarkParser.php]: Oops ! Missed that one...

File: src/NetscapeBookmarkParser.php (hunk 17 above)
Comment (TangoMan75) [reply src/NetscapeBookmarkParser.php:111]: Well in that case, shouldn't the client use  `decoder->decode()` directly ?

File: src/NetscapeBookmarkParser.php (hunk 19 above)
Comment (TangoMan75) [reply src/NetscapeBookmarkParser.php:117]: Depends on the release flow... I believe it is good practice to warn user about methods that will break before bumping version. :-/

File: src/Encoder/NetscapeBookmarkEncoder.php (hunk 14 above)
Comment (ArthurHoaro) [reply src/Encoder/NetscapeBookmarkEncoder.php:60]: No that fine, didn't think of that.

File: src/NetscapeBookmarkParser.php (hunk 17 above)
Comment (ArthurHoaro) [reply src/NetscapeBookmarkParser.php:111]: The library should have a single entry point (class) with necessary public methods, even if they're only wrappers.

File: .editorconfig (hunk 1 above)
Comment (TangoMan75) [reply .editorconfig:1]: I do not see how it is useful.

File: .gitattributes (hunk 2 above)
Comment (TangoMan75) [reply .gitattributes:1]: I don't see the purpose of this file.

File: src/NetscapeBookmarkParser.php (hunk 20 above)
Comment (TangoMan75) [reply src/NetscapeBookmarkParser.php:57]: Done.

**Comment by TangoMan75** [comment]
Hi @ArthurHoaro 👋

🙏 Sorry for long pause in the development of my PR.
I had very little time to work on it until recently, thank you for your patience.

Unfortunately  I just realized new version will not be compatible with PHP versions below 7.3 because I used type hinting profusely. That would be a bummer to revert all that work since Shaarli can still use previous version of the parser, and will have to bump PHP version at some point anyway => [PHP supported-versions](https://www.php.net/supported-versions.php) 👀

File: src/DataFormatter/BookmarksDataFormatter.php (hunk 8 above)
Comment (TangoMan75) [reply src/DataFormatter/BookmarksDataFormatter.php:59]: Every bookmark with a missing tag will own a default "undefined" tag now.

File: src/NetscapeBookmarkParser.php (hunk 19 above)
Comment (TangoMan75) [reply src/NetscapeBookmarkParser.php:117]: Oh ! And, I assumed the parser will be removed at some point. Future implementation should use Encoder and Decoder directly which makes more sense IMHO.

File: src/NetscapeBookmarkParser.php (hunk 15 above)
Comment (TangoMan75) [reply src/NetscapeBookmarkParser.php]: I removed all `try catch` blocks

File: src/NetscapeBookmarkParser.php (hunk 17 above)
Comment (TangoMan75) [reply src/NetscapeBookmarkParser.php:111]: I assumed that `NetscapeBookmarkParser` would be removed at some point. I believe it is better for users to implement their use case with new `encoder` and `decoder` directly. Still, I removed the deprecation warning as requested.

File: src/NetscapeBookmarkParser.php (hunk 18 above)
Comment (TangoMan75) [reply src/NetscapeBookmarkParser.php:90]: None of the `NetscapeBookmarkParser` methods are covered by unit tests. `NetscapeBookmarkParser` is only a wrapper around `DataFormatter`, `DataConverter` and  `Encoder` which are thoroughly covered by unit tests. Testing `export` method for example would only result on covering `file_put_contents` native php function, which doesn't add any value.

File: tests/Unit/Encoder/ParseChromiumBookmarksTest.php (hunk 22 above)
Comment (TangoMan75) [reply tests/Unit/Encoder/ParseChromiumBookmarksTest.php:306]: I had to write a small `SublimeText` python custom plugin to convert `var_dump()` in the terminal into proper php arrays... That was actually quite fun. Amazing what can be done with regular expressions.

File: phpunit.xml.dist (hunk 21 above)
Comment (TangoMan75) [reply phpunit.xml.dist:11]: Also I added a command in `Makefile` to output the coverage report, and updated `.gitignore` to exclude `coverage` folder.

**Comment by TangoMan75** [comment]
Hi @ArthurHoaro 👋

I had some time to re-work my PR :

I reverted changes to minimum supported php version, and PHPUnit version : Parser will stay compatible with PHP 7.1
Unfortunately PHPUnit coverage feature will not be available because of it though, maybe we can roll it on future release.

Well, I guess this PR is ready to merge, I hope that you find my contribution useful... 

Let me know if I missed anything ✌

**Comment by TangoMan75** [comment]
Hey @ArthurHoaro ✋

I didn't get news from you for a long time, I hope you're OK

**Comment by ArthurHoaro** [comment]
Hi @TangoMan75! I'm well, but kinda took a break from Shaarli and open source development. Sorry about that. 
I'm trying to get back into it now, so I'll run a few tests and merge this, finally!

File: .editorconfig (hunk 1 above)
Comment (ArthurHoaro) [reply .editorconfig:1]: This is useful to maintain certain coding style rules across any IDE and OS, especially for line and endings. I'd rather have it back.

**Comment by ArthurHoaro** [comment]
OK well, I didn't re-review every single line of code, but:

  - everything seems well design and commented
  - I tested an export on a large-ish file in both previous and new version and besides the key name changes, everything seems to be the same and running smoothly.
  - All the tests that we had in place seem to have been preserve and ported, which should cover almost all test cases

So let's roll!

**Review by ArthurHoaro — APPROVED** [review]

**Comment by TangoMan75** [comment]
WOW ! That's great !
